File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/listeners/ChannelListener.kt
package twilio.flutter.twilio_programmable_chat.listeners

import com.twilio.chat.Channel
import com.twilio.chat.ChannelListener as TwilioChannelListener
import com.twilio.chat.ErrorInfo
import com.twilio.chat.Member
import com.twilio.chat.Message
import twilio.flutter.twilio_programmable_chat.Mapper
import twilio.flutter.twilio_programmable_chat.TwilioProgrammableChatPlugin

class ChannelListener(private val channelSid: String) : TwilioChannelListener {
    override fun onMessageAdded(message: Message) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMessageAdded => messageSid = ${message.sid}")
        sendEvent("messageAdded", mapOf(
                "channelSid" to channelSid,
                "message" to Mapper.messageToMap(message))
            )
    }

    override fun onMessageUpdated(message: Message, reason: Message.UpdateReason) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMessageUpdated => messageSid = ${message.sid}, reason = $reason")
        sendEvent("messageUpdated", mapOf(
                "channelSid" to channelSid,
                "message" to Mapper.messageToMap(message),
                "reason" to mapOf(
                        "type" to "message",
                        "value" to reason.toString()
                )
        ))
    }

    override fun onMessageDeleted(message: Message) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMessageDeleted => messageSid = ${message.sid}")
        sendEvent("messageDeleted", mapOf(
                "channelSid" to channelSid,
                "message" to Mapper.messageToMap(message))
            )
    }

    override fun onMemberAdded(member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMemberAdded => memberSid = ${member.sid}")
        sendEvent("memberAdded", mapOf(
                "channelSid" to channelSid,
                "member" to Mapper.memberToMap(member))
            )
    }

    override fun onMemberUpdated(member: Member, reason: Member.UpdateReason) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMemberUpdated => memberSid = ${member.sid}, reason = $reason")
        sendEvent("memberUpdated", mapOf(
                "channelSid" to channelSid,
                "member" to Mapper.memberToMap(member),
                "reason" to mapOf(
                        "type" to "member",
                        "value" to reason.toString()
                )
        ))
    }

    override fun onMemberDeleted(member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMemberDeleted => memberSid = ${member.sid}")
        sendEvent("memberDeleted", mapOf(
            "channelSid" to channelSid,
            "member" to Mapper.memberToMap(member))
        )
    }

    override fun onTypingStarted(channel: Channel, member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onTypingStarted => channelSid = ${channel.sid}, memberSid = ${member.sid}")
        sendEvent("typingStarted", mapOf(
            "channelSid" to channelSid,
            "channel" to Mapper.channelToMap(channel),
            "member" to Mapper.memberToMap(member))
        )
    }

    override fun onTypingEnded(channel: Channel, member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onTypingEnded => channelSid = ${channel.sid}, memberSid = ${member.sid}")
        sendEvent("typingEnded", mapOf(
            "channelSid" to channelSid,
            "channel" to Mapper.channelToMap(channel),
            "member" to Mapper.memberToMap(member))
        )
    }

    override fun onSynchronizationChanged(channel: Channel) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onSynchronizationChanged => channelSid = ${channel.sid}")
        sendEvent("synchronizationChanged", mapOf(
            "channelSid" to channelSid,
            "channel" to Mapper.channelToMap(channel))
        )
    }

    private fun sendEvent(name: String, data: Any?, e: ErrorInfo? = null) {
        val eventData = mapOf("name" to name, "data" to data, "error" to Mapper.errorInfoToMap(e))
        TwilioProgrammableChatPlugin.channelEventSink?.success(eventData)
    }
}

File: ios/Classes/Listeners/ChannelListener.swift
import Flutter
import TwilioChatClient

public class ChannelListener: NSObject, TCHChannelDelegate {
    let channelSid: String

    init(_ channelSid: String) {
        self.channelSid = channelSid
    }

    // onMessageAdded
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, messageAdded message: TCHMessage) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMessageAdded => messageSid = \(String(describing: message.sid))")
        sendEvent("messageAdded", data: [
            "channelSid": channelSid,
            "message": Mapper.messageToDict(message, channelSid: channel.sid)
        ])
    }

    // onMessageUpdated
    public func chatClient(
        _ client: TwilioChatClient, channel: TCHChannel, message: TCHMessage, updated: TCHMessageUpdate) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMessageUpdated => messageSid = \(String(describing: message.sid)), " +
            "updated = \(String(describing: updated))")
        sendEvent("messageUpdated", data: [
            "channelSid": channelSid,
            "message": Mapper.messageToDict(message, channelSid: channel.sid),
            "reason": [
                "type": "message",
                "value": Mapper.messageUpdateToString(updated)
            ]
        ])
    }

    // onMessageDeleted
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, messageDeleted message: TCHMessage) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMessageDeleted => messageSid = \(String(describing: message.sid))")
        sendEvent("messageDeleted", data: [
            "channelSid": channelSid,
            "message": Mapper.messageToDict(message, channelSid: channel.sid)
        ])
    }

    // onMemberAdded
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, memberJoined member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMemberAdded => memberSid = \(String(describing: member.sid))")
        sendEvent("memberAdded", data: [
            "channelSid": channelSid,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onMemberUpdated
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel,
                           member: TCHMember, updated: TCHMemberUpdate) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMemberUpdated => memberSid = \(String(describing: member.sid)), " +
            "updated = \(String(describing: updated))")
        sendEvent("memberUpdated", data: [
            "channelSid": channelSid,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any,
            "reason": [
                "type": "member",
                "value": Mapper.memberUpdateToString(updated)
            ]
        ])
    }

    // onMemberDeleted
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, memberLeft member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMemberDeleted => memberSid = \(String(describing: member.sid))")
        sendEvent("memberDeleted", data: [
            "channelSid": channelSid,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onTypingStarted
    public func chatClient(_ client: TwilioChatClient, typingStartedOn channel: TCHChannel, member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onTypingStarted => channelSid = \(String(describing: channel.sid)), " +
            "memberSid = \(String(describing: member.sid))")
        sendEvent("typingStarted", data: [
            "channelSid": channelSid,
            "channel": Mapper.channelToDict(channel) as Any,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onTypingEnded
    public func chatClient(_ client: TwilioChatClient, typingEndedOn channel: TCHChannel, member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onTypingEnded => channelSid = \(String(describing: channel.sid)), " +
            "memberSid = \(String(describing: member.sid))")
        sendEvent("typingEnded", data: [
            "channelSid": channelSid,
            "channel": Mapper.channelToDict(channel) as Any,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onSynchronizationChanged
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel,
                           synchronizationStatusUpdated status: TCHChannelSynchronizationStatus) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onSynchronizationChanged => channelSid = \(String(describing: channel.sid))")
        sendEvent("synchronizationChanged", data: [
            "channelSid": channelSid,
            "channel": Mapper.channelToDict(channel) as Any
        ])
    }

    private func sendEvent(_ name: String, data: [String: Any]? = nil, error: Error? = nil) {
        let eventData = [
            "name": name,
            "data": data,
            "error": Mapper.errorToDict(error)
            ] as [String: Any?]

        if let events = SwiftTwilioProgrammableChatPlugin.channelEventSink {
            events(eventData)
        }
    }
}
